python/sanlock.c

```diff
@@ -1688,75 +1690,93 @@ initexception(void)
-    PYSNLK_INIT_ADD_CONSTANT(SANLK_HOST_LIVE, "HOST_LIVE");
-    PYSNLK_INIT_ADD_CONSTANT(SANLK_HOST_FAIL, "HOST_FAIL");
-    PYSNLK_INIT_ADD_CONSTANT(SANLK_HOST_DEAD, "HOST_DEAD");
-    PYSNLK_INIT_ADD_CONSTANT(SANLK_HOST_UNKNOWN, "HOST_UNKNOWN");
+    if (PyModule_AddIntConstant(m, "HOST_FREE", SANLK_HOST_FREE))
+        return -1;
+    if (PyModule_AddIntConstant(m, "HOST_LIVE", SANLK_HOST_LIVE))
+        return -1;
+    if (PyModule_AddIntConstant(m, "HOST_FAIL", SANLK_HOST_FAIL))
+        return -1;
+    if (PyModule_AddIntConstant(m, "HOST_DEAD", SANLK_HOST_DEAD))
+        return -1;
+    if (PyModule_AddIntConstant(m, "HOST_UNKNOWN", SANLK_HOST_UNKNOWN))
+        return -1;
 
     /* set event flags */
-    PYSNLK_INIT_ADD_CONSTANT(SANLK_SETEV_CUR_GENERATION, "SETEV_CUR_GENERATION");
-    PYSNLK_INIT_ADD_CONSTANT(SANLK_SETEV_CLEAR_HOSTID,   "SETEV_CLEAR_HOSTID");
-    PYSNLK_INIT_ADD_CONSTANT(SANLK_SETEV_CLEAR_EVENT,    "SETEV_CLEAR_EVENT");
-    PYSNLK_INIT_ADD_CONSTANT(SANLK_SETEV_REPLACE_EVENT,  "SETEV_REPLACE_EVENT");
-    PYSNLK_INIT_ADD_CONSTANT(SANLK_SETEV_ALL_HOSTS,      "SETEV_ALL_HOSTS");
-
-#undef PYSNLK_INIT_ADD_CONSTANT
+    if (PyModule_AddIntConstant(m, "SETEV_CUR_GENERATION", SANLK_SETEV_CUR_GENERATION))
+        return -1;
+    if (PyModule_AddIntConstant(m, "SETEV_CLEAR_HOSTID", SANLK_SETEV_CLEAR_HOSTID))
+        return -1;
+    if (PyModule_AddIntConstant(m, "SETEV_CLEAR_EVENT", SANLK_SETEV_CLEAR_EVENT))
+        return -1;
+    if (PyModule_AddIntConstant(m, "SETEV_REPLACE_EVENT", SANLK_SETEV_REPLACE_EVENT))
+        return -1;
+    if (PyModule_AddIntConstant(m, "SETEV_ALL_HOSTS", SANLK_SETEV_ALL_HOSTS))
+        return -1;
 
     /* Tuples with supported sector size and alignment values */
     PyObject *sector = Py_BuildValue("ii", SECTOR_SIZE_512, SECTOR_SIZE_4K);
     if (!sector)
-        return;
-    if (PyModule_AddObject(py_module, "SECTOR_SIZE", sector)) {
+        return -1;
+    if (PyModule_AddObject(m, "SECTOR_SIZE", sector)) {
         Py_DECREF(sector);
-        return;
+        return -1;
     }
 
-    PyObject *align = Py_BuildValue("llll", ALIGNMENT_1M, ALIGNMENT_2M, ALIGNMENT_4M, ALIGNMENT_8M);
+    PyObject *align = Py_BuildValue(
+            "llll", ALIGNMENT_1M, ALIGNMENT_2M, ALIGNMENT_4M, ALIGNMENT_8M);
     if (!align)
-      return;
-    if (PyModule_AddObject(py_module, "ALIGN_SIZE", align)) {
+        return -1;
+    if (PyModule_AddObject(m, "ALIGN_SIZE", align)) {
         Py_DECREF(align);
-	return;
+        return -1;
+    }
+
+    return 0;
+}
+
+PyMODINIT_FUNC
+initsanlock(void)
+{
+    PyObject* m = Py_InitModule4(
+        MODULE_NAME,
+        sanlock_methods,
+        pydoc_sanlock,
+        NULL,
+        PYTHON_API_VERSION);
+
+    if (m == NULL) {
+        PyErr_Format(PyExc_ImportError, "Failed to allocate module: %s", MODULE_NAME);
+        return;
     }
 
+    if(module_init(m))
+        PyErr_Format(PyExc_ImportError, "Failed to initialize module: %s", MODULE_NAME);
```

Comment:
If module_init() failed, it already set the error, or a function it called set the error (e.g. MemoryError). This error will hide the real cause of the failure.

We don't have any way to signal errors in python 2 (except setting an error), so we cannot do anything useful when module_init fail, except returning.

Maybe add a comment about that to save time for the next one looking this code.